packages/graphql/src/services/learningElements.ts

```diff
@@ -194,161 +193,155 @@ export async function respondToQuestionInstance(
+      if (!instance) {
+        return {}
       }
-    }
 
-    const questionData =
-      instance?.questionData?.valueOf() as AllQuestionTypeData
-    const results = instance?.results?.valueOf() as AllQuestionResults
+      // if the participant had already responded, don't track the new response
+      // keeps the evaluation more accurate, as repeated entries do not skew into the "correct direction"
+      const hasPreviousResponse = instance?.responses.length > 0
+      if (ctx.user?.sub && !treatAnonymous && hasPreviousResponse) {
+        return {
+          instance,
+          updatedInstance: instance,
+        }
+      }
 
-    if (!questionData) {
-      return {}
-    }
+      const questionData = instance?.questionData
+      const results = instance?.results
 
-    let updatedResults:
-      | {
-          choices?: Record<string, number>
-        }
-      | Record<string, number> = {}
-
-    switch (questionData.type) {
-      case QuestionType.SC:
-      case QuestionType.MC:
-      case QuestionType.KPRIM: {
-        updatedResults.choices = response.choices!.reduce(
-          (acc, ix) => ({
-            ...acc,
-            [ix]: acc[ix] + 1,
-          }),
-          results.choices as Record<string, number>
-        )
-        break
+      if (!questionData) {
+        return {}
       }
 
-      case QuestionType.NUMERICAL: {
-        if (
-          typeof response.value === 'undefined' ||
-          response.value === null ||
-          response.value === ''
-        ) {
-          return {}
+      let updatedResults:
+        | {
+            choices?: Record<string, number>
+          }
+        | Record<string, number> = {}
+
+      switch (questionData.type) {
+        case QuestionType.SC:
+        case QuestionType.MC:
+        case QuestionType.KPRIM: {
+          updatedResults.choices = response.choices!.reduce(
+            (acc, ix) => ({
+              ...acc,
+              [ix]: acc[ix] + 1,
+            }),
+            results.choices as Record<string, number>
+          )
+          break
         }
 
-        const parsedValue = parseFloat(response.value)
-        if (
-          isNaN(parsedValue) ||
-          (typeof questionData.options.restrictions!.min === 'number' &&
-            parsedValue < questionData.options.restrictions!.min) ||
-          (typeof questionData.options.restrictions!.max === 'number' &&
-            parsedValue > questionData.options.restrictions!.max)
-        ) {
-          return {}
-        }
+        case QuestionType.NUMERICAL: {
+          if (
+            typeof response.value === 'undefined' ||
+            response.value === null ||
+            response.value === ''
+          ) {
+            return {}
+          }
 
-        const value = String(parsedValue)
+          const parsedValue = parseFloat(response.value)
+          if (
+            isNaN(parsedValue) ||
+            (typeof questionData.options.restrictions!.min === 'number' &&
```

Comment:
TS correctly recognizes the questionData.options object here as `QuestionOptionsNumerical`. The restrictions are optional on this type right now, but still it is assumed that they exist. Should we either assume that the restriction could possibly not exist as suggested by the type or enforce on the type that numerical options have restrictions, even though min and/or max on them might not be defined?

Reply:
changed to ?